qp_klp/klp.py

```diff
@@ -158,3 +161,138 @@ def sequence_processing_pipeline(qclient, job_id, parameters, out_dir):
+    sheet.Header['SheetType'] = 'standard_metag'
+    sheet.Header['SheetVersion'] = '90'
+    sheet.Header['Investigator Name'] = 'Qiita'
+    sheet.Header['Experiment Name'] = project_name
+
+    sheet.Bioinformatics = pd.DataFrame(
+        columns=['Sample_Project', 'ForwardAdapter', 'ReverseAdapter',
+                 'library_construction_protocol',
+                 'experiment_design_description',
+                 'PolyGTrimming', 'HumanFiltering', 'QiitaID'],
+        data=[[project_name, 'NA', 'NA', 'NA', 'NA', 'FALSE', 'TRUE', sid]])
+
+    for k, vals in pt.iterrows():
+        k = k.split('.', 1)[-1]
+        sample = {
+            'Sample_Name': k,
+            'Sample_ID': k.replace('.', '_'),
+            'Sample_Plate': '',
+            'well_id_384': '',
+            'I7_Index_ID': '',
+            'index': vals['index'],
+            'I5_Index_ID': '',
+            'index2': vals['index2'],
+            'Sample_Project': project_name,
+            'Well_description': '',
+            'Sample_Well': '',
+            'Lane': '1'}
+        sheet.add_sample(sample_sheet.Sample(sample))
+
+    sheet.Contact = pd.DataFrame(
+        columns=['Email', 'Sample_Project'],
+        data=[['qiita.help@gmail.com', project_name]])
+
+    new_sample_sheet = out_path('sample-sheet.csv')
+    with open(new_sample_sheet, 'w') as f:
+        sheet.write(f, 1)
+
+    # now that we have a sample_sheet we can fake the
+    # ConvertJob folder so we are ready for the restart
+    convert_path = out_path('ConvertJob')
+    project_folder = out_path('ConvertJob', project_name)
+    makedirs(project_folder, exist_ok=True)
+
+    for _, fs in files.items():
+        for f in fs:
+            bn = basename(f['filepath'])
+            symlink(f['filepath'], f'{project_folder}/{bn}')
+
+    # create job_completed file to skip this step
+    with open(f'{convert_path}/job_completed', 'w') as f:
+        f.write('')
+
+    try:
+        kwargs = {'qclient': qclient,
+                  'uif_path': new_sample_sheet,
+                  'lane_number': "1",
+                  'config_fp': CONFIG_FP,
+                  'run_identifier': '211021_A00000_0000_SAMPLE',
+                  'output_dir': out_dir,
+                  'job_id': job_id,
+                  'status_update_callback': status_line.update_job_status,
+                  # set 'update_qiita' to False to avoid updating Qiita DB
+                  # and copying files into uploads dir. Useful for testing.
+                  'update_qiita': True,
+                  'is_restart': True}
+
+        workflow = WorkflowFactory().generate_workflow(**kwargs)
+
+        status_line.update_job_status("Getting project information")
+
+        workflow.execute_pipeline()
+
+        status_line.update_job_status("SPP finished")
+
+    except (PipelineError, WorkflowError) as e:
+        # assume AttributeErrors are issues w/bad sample-sheets or
+        # mapping-files.
+        return False, None, str(e)
```

Comment:
The traceback is lost here:

```
In [4]: def foo():
   ...:     raise ValueError("stuff")
   ...: 

In [5]: try:
   ...:     foo()
   ...: except ValueError as e:
   ...:     print(str(e))
   ...: 
stuff
```

See this [SO](https://stackoverflow.com/a/3702847) post. I think the intent is `str(e)` to instead be `traceback.format_exc()`, right? If so then `import traceback` is also needed

Reply:
AFAIK, the intent is completely the oposite, just raise/report the error in the GUI with minimal information for the user. I think that way, if is something obvious and handled, like "sample sheet has wrong OverwriteCycles value", that's what's shown but if there is something less obvious, users will need to contact the admins/devs to investigate. FWIW, this has been a useful interaction for this specific plugin: wet/dry-lab interactions.

Reply:
How do the dev's know what line in the codebase is throwing the exception without the traceback?

Reply:
But the developer will not know the exact line of code raising the exception. Won't that require the developer to then guess or perform a much more time expensive debugging process to determine what specifically failed?

Reply:
I see what you are saying but in experience so far that's not the case. However, we might be missing something so I'll change and we can revert back if users get too annoyed.

Reply:
thanks! For users, the traceback could either be post processed, or an additional item in the tuple could be returned (the original `str(e)`)

Reply:
Yeah, decided to write a log in the outdir so devs can see the full traceback but keep it simple ("str(e)") for users.